core/src/main/java/org/apache/calcite/sql/SqlSnapshot.java

```diff
@@ -127,10 +129,25 @@ private SqlSnapshotOperator() {
         int leftPrec,
         int rightPrec) {
       final SqlSnapshot snapshot = (SqlSnapshot) call;
-
-      snapshot.tableRef.unparse(writer, 0, 0);
-      writer.keyword("FOR SYSTEM_TIME AS OF");
-      snapshot.period.unparse(writer, 0, 0);
+      SqlNode tableRef = snapshot.tableRef;
+
+      if (tableRef instanceof SqlBasicCall
+          && ((SqlBasicCall) tableRef).getOperator() instanceof SqlAsOperator) {
+        SqlBasicCall basicCall = (SqlBasicCall) tableRef;
+        basicCall.operand(0).unparse(writer, 0, 0);
+        writer.setNeedWhitespace(true);
+        writeForSystemTimeAsOf(writer, snapshot);
+        writer.keyword("AS");
+        basicCall.operand(1).unparse(writer, 0, 0);
+      } else {
+        tableRef.unparse(writer, 0, 0);
+        writeForSystemTimeAsOf(writer, snapshot);
+      }
     }
   }
+
+  private static void writeForSystemTimeAsOf(SqlWriter writer, SqlSnapshot snapshot) {
+    writer.keyword("FOR SYSTEM_TIME AS OF");
+    snapshot.period.unparse(writer, 0, 0);
```

Comment:
Maybae it is better to put `writeForSystemTimeAsOf` in SqlSnapshotOperator.

Reply:
ok,I'll change it